### PR TITLE
feat: add encryption service

### PR DESCRIPTION
**Pull request by djabarovgeorge**

### CODE REVIEW

**Comment by djabarovgeorge** [comment]
At the moment I added aes-256 algorithm with iv in order to add some security due to the reason the web like it, but still did not have the time to dive deeper into the pros and cons of this algorithm, will dive into it later again.

Will love to hear any opinions if anyone is familiar with good algorithms to store credentials on the DB.

File: apps/api/src/app/shared/services/encryption/encrypt-provider.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { decrypt, encrypt } from './cipher';
+import { ICredentialsDto, secureCredentials } from '@novu/shared';
+
+const novuSubMask = 'nvsk.';
+
+export function encryptProviderSecret(text) {
+  const encrypted = encrypt(text);
+
+  return novuSubMask + encrypted;
```
Comment (scopsy) [inline apps/api/src/app/shared/services/encryption/encrypt-provider.ts:9]: Can we add types here?

File: apps/api/src/app/shared/services/encryption/cipher.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  const textParts = text.split(':');
+  const iv = Buffer.from(textParts.shift(), 'hex');
+  const encryptedText = Buffer.from(textParts.join(':'), 'hex');
+  const decipher = createDecipheriv('aes-256-cbc', Buffer.from(ENCRYPTION_KEY), iv);
+  let decrypted = decipher.update(encryptedText);
+
+  decrypted = Buffer.concat([decrypted, decipher.final()]);
+
+  return decrypted.toString();
+}
```
Comment (scopsy) [inline apps/api/src/app/shared/services/encryption/cipher.ts:26]: Would be great to have unit tests and typescript annotations here aswell 🙏

File: apps/api/src/app/shared/services/encryption/cipher.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { createCipheriv, createDecipheriv, randomBytes } from 'crypto';
+
+const ENCRYPTION_KEY = process.env.P_ENCRYPTION_KEY;
```
Comment (scopsy) [inline apps/api/src/app/shared/services/encryption/cipher.ts:3]:
```suggestion
const ENCRYPTION_KEY = process.env.STORE_ENCRYPTION_KEY;
```

File: apps/api/src/app/shared/services/encryption/encrypt-provider.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+export function decryptCredentials(credentials: ICredentialsDto): ICredentialsDto {
+  const decryptedCredentials: ICredentialsDto = {};
+
+  Object.keys(credentials).forEach((key) => {
+    if (credentials[key].startsWith(novuSubMask)) {
+      decryptedCredentials[key] = decryptProviderSecret(credentials[key]);
+    } else {
+      decryptedCredentials[key] = credentials[key];
+    }
+  });
```
Comment (scopsy) [inline apps/api/src/app/shared/services/encryption/encrypt-provider.ts:43]: Is t here a chance here of nested objects?

**Review by scopsy — COMMENTED** [review]
Looks great @djabarovgeorge added a few comments. The main thing is that I'm missing here unit tests for this functionality and also some typescript annotations. Where did you took the examples for the encryption and decryption logic?

File: apps/api/src/app/shared/services/encryption/encrypt-provider.ts (hunk 1 above)
Comment (djabarovgeorge) [reply apps/api/src/app/shared/services/encryption/encrypt-provider.ts:9]: Sure thing I will add shortly!

File: apps/api/src/app/shared/services/encryption/cipher.ts (hunk 2 above)
Comment (djabarovgeorge) [reply apps/api/src/app/shared/services/encryption/cipher.ts:26]: Agree i really wanted to add a unit test here but missed the rush of continuing, however, I did add an e2e test on the new usecase I created on the next PR but the unit test will be much more appropriate here I will look into it.! :)

File: apps/api/src/app/shared/services/encryption/cipher.ts (hunk 3 above)
Comment (djabarovgeorge) [reply apps/api/src/app/shared/services/encryption/cipher.ts:3]: Haha agree i look at it today and did not like the name as well and changed to other one, will update to 'store' :)))

File: apps/api/src/app/shared/services/encryption/encrypt-provider.ts (hunk 4 above)
Comment (djabarovgeorge) [reply apps/api/src/app/shared/services/encryption/encrypt-provider.ts:43]: No there is not at the moment the credentials object is only 'list' of string props.

**Comment by djabarovgeorge** [comment]
@scopsy  Oh almost forgot the script was inspired from following gist https://gist.github.com/vlucas/2bd40f62d20c1d49237a109d491974eb

File: apps/api/src/.example.env [hunk 5]
```diff
@@ -2,6 +2,7 @@ NODE_ENV=local
 PORT=3000
 API_ROOT_URL=http://localhost:3000
 FRONT_BASE_URL=http://localhost:4200
+STORE_ENCRYPTION_KEY="<ENCRYPTION_KEY_MUST_BE_32_LONG>"
```
Comment (scopsy) [inline apps/api/src/.example.env:5]: Could we also add this to the `docker` folder example and pass it to the docker-compose file from the .env file?

**Review by scopsy — APPROVED** [review]

**Pull request merged**